### ISSUE TITLE
twisted.python.logfile.LogFile's defaultMode parameter is undocumented, untested, insecure

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2586]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2586|
|Type|defect|
|Created|2007-04-20 13:42:15Z|

Undocumented and untested speak for themselves.  Regarding insecure, the defaultMode should be used when creating new files.  It shouldn't be applied _after_ the file is created, since that creates a window where the __actual__ default mode might allow an attacker to open the file.

<details><summary>Searchable metadata</summary>

```
trac-id__2586 2586
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1177076535000000 1177076535000000
changetime__1178525904000000 1178525904000000
version__None None
owner__ 
cc__therve cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

I'm ready to do it after [#2575](https://github.com/twisted/twisted/issues/2575).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve removed owner|
|-|-|

Ready to review in logfile-defaultmode-2586.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @therve|
|-|-|

Use parens for breaking long lines:

```
       if (defaultMode is None and os.path.exists(self.path) and
           hasattr(os, "chmod")):
           self.defaultMode = stat.S_IMODE(os.stat(self.path)[stat.ST_MODE])
```

Although, what version of Python doesn't have os.chmod?  Is there any test coverage for correctly handling the case where it's missing?  Maybe it should be dropped.

The umask mangling code needs some try/finally action in case opening the file fails.  Also we should open the file with ```file``` instead of ```open``` (because we hate Python).

It should be possible to set the logfiles permission using only a umask call.  This is only slightly better than doing it with a umask and a chmod call (one fewer syscall), so I don't really care which way it gets done, just thought I'd point out that it could be shorter.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @exarkun|
|-|-|

I removed the compatibility code, there was already tests that use it. I also wrap umask with a try/finally.

Why do you prefer file over open?

Also, I don't think it's possible to only use umask. It's only the mask used after open, and I think default mode for open is 666, so you'll never get the executable bit with umask.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @therve|
|-|-|

Guido keeps promising to make open a factory function with some extra behavior, whereas since file is the type object for actual file objects, it will keep its current behavior.

Oops, execute bit didn't occur to me.  I can't think of a reason to have executable log files, but there's no reason to disallow it either.  I withdraw that comment.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve removed owner|
|-|-|

Okay, done for file.

Regarding executable log files, I have lots of awesome usecases, but I don't think you want to know (just kidding).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve|
|-|-|

Looks good, please merge :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set status to `closed`|
|-|-|

(In [20150]) Merge logfile-defaultmode-2586

Author: therve
Reviewer: exarkun
Fixes [#2586](https://github.com/twisted/twisted/issues/2586)

Add documentation and tests to the defaultMode arguments of LogFile, and fix
a potential security problem when using custom mode.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve commented|
|-|-|

(In [20151]) Revert [20150], regressions in windows slaves

Refs [#2586](https://github.com/twisted/twisted/issues/2586)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve removed owner<br>@therve set status to `new`|
|-|-|

Buildbot seems happier now.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_15"></a>|@ralphm set owner to @therve|
|-|-|

Looks good to me. Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve set status to `closed`|
|-|-|

(In [20199]) Merge logfile-defaultmode-2586

Author: therve
Reviewer: ralphm, exarkun
Fixes [#2586](https://github.com/twisted/twisted/issues/2586)

Add documentation and tests to the defaultMode arguments of LogFile,
and fix a potential security problem when using custom mode.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
